pkg/csi/service/common/commonco/k8sorchestrator/wcp_node.go

```diff
@@ -0,0 +1,107 @@
+
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"sigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/common"
+	"sigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/logger"
+)
+
+var clusterComputeResourceMoIds []string
+
+// InitializeCSINodes listens on node Add events to create CSINode instance for each K8s node object.
+func (c *K8sOrchestrator) InitializeCSINodes(ctx context.Context) error {
+	log := logger.GetLogger(ctx)
+	var err error
+	clusterComputeResourceMoIds, err = common.GetClusterComputeResourceMoIds(ctx)
+	if err != nil {
+		return logger.LogNewErrorf(log, "failed to get clusterComputeResourceMoIds. Error: %v", err)
+	}
+	c.informerManager.AddNodeListener(
+		func(obj interface{}) { // Add.
+			c.createCSINode(obj)
+		},
+		nil,
+		nil)
+	log.Infof("Starting informer to listen on nodes to create CSINodes")
+	return nil
+}
+
+// createCSINode creates CSINode instances for each K8s node object.
+func (c *K8sOrchestrator) createCSINode(obj interface{}) {
+	ctx, log := logger.GetNewContextWithLogger()
+	node, ok := obj.(*corev1.Node)
+	if node == nil || !ok {
+		log.Warnf("createCSINode: unrecognized object %+v found", obj)
+		return
+	}
+
+	// Check if CreateCSINodeAnnotation is present on the node. Spherelet adds this annotation on
+	// the node to communicate to CSI driver to create CSINode instance for each node.
+	csiNodeAnnotation, exists := node.ObjectMeta.Annotations[common.CreateCSINodeAnnotation]
+	if val, _ := strconv.ParseBool(csiNodeAnnotation); !exists || !val {
+		log.Infof("createCSINode: %s annotation exists: %t with value: %t on the node %s",
+			common.CreateCSINodeAnnotation, exists, val, node.Name)
+		return
+	}
+
+	var topologyKeysList []string
+	if len(clusterComputeResourceMoIds) > 1 {
+		// Publish zone and host standard topology keys for stretch supervisor.
+		topologyKeysList = append(topologyKeysList, corev1.LabelTopologyZone, corev1.LabelHostname)
+	} else {
+		topologyKeysList = append(topologyKeysList, corev1.LabelHostname)
+	}
+	// NOTE: As this is a WCP node, we can safely assume that only vSphere CSI driver will be run on it.
+	// If spherelet is not creating the CSINodes, we will create them and ignore the error if it is already present.
+	csiNodeSpec := &storagev1.CSINode{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: string(node.Name),
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "v1",
+					Kind:       "Node",
+					Name:       node.Name,
+					UID:        node.UID,
+				},
+			},
+		},
+		Spec: storagev1.CSINodeSpec{
+			Drivers: []storagev1.CSINodeDriver{
+				{
+					Name:         common.VSphereCSIDriverName,
+					NodeID:       node.Name,
+					TopologyKeys: topologyKeysList,
+				},
+			},
+		},
+	}
+	csinode, err := c.k8sClient.StorageV1().CSINodes().Create(ctx, csiNodeSpec, metav1.CreateOptions{})
+	if err != nil {
+		log.Errorf("failed to create CSINode object %+v. Error: %v", csiNodeSpec, err)
```

Comment:
if driver fails to create CSINode object here then when it will try again?

Reply:
If the error is not `AlreadyExists`, we crash the syncer container so that we can retry creation of CSINode objects.